### ISSUE TITLE
`RS_INTERNAL_ERROR` (job made no progress) when when `RS_BLOCKED` and `*len = 0` is returned from `rs_copy_cb`

### ISSUE THREAD

**Issue by astro-stan** [librsync/librsync#258]
Currently, in `rs_job_iter`, it is checked if the result from `rs_job_work` is `RS_DONE` or `RS_BLOCKED`. If it is, and no data was moved in /out of the job buffers (`rs_buffers_t`), an `RS_INTERNAL_ERROR` is returned, along with an error message that the job made no progress. Relevant code:

https://github.com/librsync/librsync/blob/ee3df5c8775d571871170c52613c36af1a51db36/src/job.c#L93-L104

I can see the logic behind this, however I have the following use case:

```c
rs_result my_file_copy_cb(void * arg, rs_long_t pos, size_t * len, void ** buf) {
  ...

  /* Oh, I need to wait for the data to become available */
  if (not_ready_yet) {
    /* Lets return blocked result, so that I do not block  the thread running `rs_job_iter` */
    *len = 0;
    return RS_BLOCKED;
  }
}

// In another file
void myProgramLoop() {
  ...
  job = rs_patch_begin(my_file_copy_cb, ...);

  while (1) {
    /* Do some other stuff */
    ...
    result = rs_job_iter(job, ...); /* result = RS_INTERNAL_ERROR */

    if (result != RS_DONE || result != RS_BLOCKED) {
      /* PANIC! */
    }
    ...
  }
}
```

It seems like the `rs_job_iter` does not take into consideration that I/O blockage can also occur while trying to fetch the basis for the patch.

**Comment by dbaarda** [comment]
I agree this probably needs a little bit of a re-think, but I'm not entirely sure it's wrong.

However, notice that this error is only triggered if there is input and output data available and no progress is made. While it is true that fetching more input and writing output can be I/O blocked, I believe `rs_job_iter()` should be able to suck all available input into an internal buffer even if it can't make progress otherwise. This means it should be able to empty the `buffers->avail_in` which would prevent this from happening.

At least I think so. I can't remember if the internal buffer is also limited in size, so maybe it can happen that it can't empty the input buffer?

**Comment by dbaarda** [comment]
Ah... hang on, this is in patch, not delta. The delta operation is way more complicated so that's what I spent more time debugging and making sure things like this can't happen. The patch is pretty simple, but it's also quite possible it does have a silly bug like this that no-one has stumbled into yet.

**Comment by astro-stan** [comment]
Thanks for responding.

Yeah, I think this should be considered a bug. What do you think is the best way to fix this?

**Comment by dbaarda** [comment]
An MR would be much appreciated since I don't have much time to tackle this myself right now. I can probably give some guidance though.
